contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+        MultiHopFillData memory fillData,
+        uint256 minBuyAmount
+    )
+        public
+        payable
+        override
+        returns (uint256 outputTokenAmount)
+    {
+        require(
+            fillData.tokens.length == fillData.calls.length + 1,
+            "MultiplexFeature::multiHopFill/MISMATCHED_ARRAY_LENGTHS"
+        );
+        IERC20TokenV06 outputToken = IERC20TokenV06(fillData.tokens[fillData.tokens.length - 1]);
+        uint256 senderBalanceBefore = outputToken.getTokenBalanceOf(msg.sender);
+        uint256 ethBalanceBefore = address(this).balance.safeSub(msg.value);
+
+        _multiHopFill(fillData, msg.value);
+
+        // RFC: I think these balance checks are arguably unnecessary,
+        // see reasoning above.
+
+        // The `outputTokenAmount` returned by `_multiHopFill` may not
+        // be fully accurate (e.g. due to some janky token and/or
+        // reentrancy situation).
+        outputTokenAmount = outputToken.getTokenBalanceOf(msg.sender)
+            .safeSub(senderBalanceBefore);
+        require(
+            outputTokenAmount >= minBuyAmount,
+            "MultiplexFeature::multiHopFill/UNDERBOUGHT"
+        );
+
+        uint256 ethBalanceAfter = address(this).balance;
+        require(
+            ethBalanceAfter >= ethBalanceBefore,
+            "MultiplexFeature::multiHopFill/OVERSPENT_ETH"
+        );
+        // Refund ETH
+        if (ethBalanceAfter > ethBalanceBefore) {
+            payable(msg.sender).transfer(ethBalanceAfter - ethBalanceBefore);
+        }
+    }
+
+    // Similar to FQT. If `fillData.sellAmount` is set to `type(uint256).max`,
+    // this is effectively a batch fill. Otherwise it can be set to perform a
+    // market sell of some amount. In the case of a market sell, the `sellAmount`
+    // values of the constituent wrapped calls can encode percentages of the
+    // full `fillData.sellAmount`. This is useful if e.g. the `_batchFill` call
+    // the second leg of a multi-hop fill, where the sell amount is not known
+    // ahead of time.
+    function _batchFill(BatchFillData memory fillData, uint256 totalEth)
+        public
+        returns (uint256 outputTokenAmount, uint256 remainingEth)
+    {
+        // Track the remaining ETH allocated to this call.
+        remainingEth = totalEth;
+        // Track the amount of input token sold.
+        uint256 soldAmount;
+        for (uint256 i = 0; i != fillData.calls.length; i++) {
+            // Check if we've hit our target.
+            if (soldAmount >= fillData.sellAmount) { break; }
+            WrappedBatchCall memory wrappedCall = fillData.calls[i];
+            // Compute the fill amount.
+            uint256 inputTokenAmount = _normalizeAmount(
+                wrappedCall.sellAmount,
+                fillData.sellAmount,
+                fillData.sellAmount.safeSub(soldAmount)
+            );
+            if (wrappedCall.selector == INativeOrdersFeature._fillRfqOrder.selector) {
```

Comment:
Nit - this is a really long if statement. It could help readability to split logic from each clause into their own function. I imagine this would not affect efficiency since it's just jump.